R/sampler-hmc.R

```diff
@@ -0,0 +1,220 @@
+      ## Make a full step for the position
+      theta_next <- theta_next + epsilon * v_next
+      pars_next <- internal$transform$rn2model(theta_next)
+      ## Make a full step for the momentum, except at end of trajectory
+      if (i != n_integration_steps) {
+        v_next <- v_next + epsilon * compute_gradient(pars_next)
+      }
+      if (debug) {
+        history[i + 1, ] <- pars_next
+      }
+    }
+
+    ## Make a half step for momentum at the end.
+    v_next <- v_next + epsilon * compute_gradient(pars_next) / 2
+
+    ## Some treatments would negate momentum at end of trajectory to
+    ## make the proposal symmetric by setting v_next <- -v_next but
+    ## this is not actually needed in practice.
+
+    ## Potential energy at the end of the trajectory
+    density_next <- model$density(pars_next)
+
+    ## Kinetic energies at the start (energy) and end (energy_next) of
+    ## the trajectories
+    energy <- sum(v^2) / 2
+    energy_next <- sum(v_next^2) / 2
+
+    ## Accept or reject the state at end of trajectory, returning
+    ## either the position at the end of the trajectory or the initial
+    ## position
+    u <- rng$random_real(1)
+    accept <- u < exp(density_next - state$density + energy - energy_next)
+    if (accept) {
+      state$density <- density_next
+      state$pars <- pars_next
+    }
+
+    if (debug) {
+      internal$history <-
+        c(internal$history, list(list(pars = history, accept = accept)))
+    }
+
+    state
+  }
+
+  finalise <- function(state, model, rng) {
+    if (debug) {
+      pars <- lapply(internal$history, "[[", "pars")
+      pars <- array(unlist(pars),
+                    c(dim(pars[[1]]), length(pars)),
+                    list(NULL, model$parameters, NULL))
+      accept <- vlapply(internal$history, "[[", "accept")
+      list(debug = list(pars = pars, accept = accept))
+    } else {
+      NULL
+    }
+  }
+
+  mcstate_sampler("Hamiltonian Monte Carlo",
+                  initialise,
+                  step,
+                  finalise)
+}
+
+
+hmc_transform <- function(domain) {
+  lower <- domain[, 1]
+  upper <- domain[, 2]
+
+  is_bounded <- is.finite(lower) & is.finite(upper)
+  is_semi_infinite <- is.finite(lower) & upper == Inf
+  is_infinite <- lower == -Inf & upper == Inf
+  unknown <- !(is_bounded | is_semi_infinite | is_infinite)
+  if (any(unknown)) {
+    ## This is not very likely to be thrown
+    cli::cli_abort("Unhandled domain type for parameter {which(unknown)}")
+  }
+
+  ## Save finite bounds for later:
+  a <- lower[is_bounded]
+  b <- upper[is_bounded]
+
+  list(
+    ## Transform from R^n into the model space
+    rn2model = function(theta) {
+      theta[is_semi_infinite] <- exp(theta[is_semi_infinite])
+      theta[is_bounded] <- ilogit_bounded(theta[is_bounded], a, b)
+      theta
+    },
+
+    ## Transform from model space to R^n
+    model2rn = function(x) {
+      x[is_semi_infinite] <- log(x[is_semi_infinite])
```

Comment:
I suggest instead `x[is_semi_infinite] <- log(x[is_semi_infinite] - lower[is_semi_infinite])`. Reasons being: 1. this ensures we transform to R^n and 2. the current version will break if you have a lower bound below 0.

Reply:
Relatedly the rn2model would be
```
theta[is_semi_infinite] <- lower[is_semi_infinite] + exp(theta[is_semi_infinite])
```

Reply:
And if we wanted to allow semi-infinite but with a finite upper bound we could use `log(upper - x)` as a transform

Reply:
thanks Ed - I've applied the semi infinite fix; I've not done the one for semi infinite in the negative direction yet though; let's wait until we need it first